core/src/models/ExportProcessor.ts

```diff
@@ -0,0 +1,225 @@
+import { api, config } from "actionhero";
+import * as uuid from "uuid";
+import { Destination } from "./Destination";
+import { APIData } from "../modules/apiData";
+import { StateMachine } from "../modules/stateMachine";
+import { Export } from "./Export";
+import { ExportProcessorOps } from "../modules/ops/exportProcessor";
+
+const ERROR_LEVELS = ["error", "info"] as const;
+export type ExportProcessorErrorLevel = typeof ERROR_LEVELS[number];
+
+export const ExportProcessorStates = [
+  "pending", // waiting to be processed
+  "failed", // something went wrong and we won't try again
+  "complete", // OK!
+] as const;
+
+const STATE_TRANSITIONS = [
+  { from: "pending", to: "failed", checks: [] },
+  {
+    from: "pending",
+    to: "complete",
+    checks: [
+      (instance: ExportProcessor) => instance.ensureAllExportsProcessed(),
+    ],
+  },
+];
+
+@Table({ tableName: "exportProcessors", paranoid: false })
+export class ExportProcessor extends Model {
+  idPrefix() {
+    return "prc";
+  }
+
+  @Column({ primaryKey: true })
+  id: string;
+
+  @CreatedAt
+  createdAt: Date;
+
+  @UpdatedAt
+  updatedAt: Date;
+
+  @AllowNull(false)
+  @ForeignKey(() => Destination)
+  @Column
+  destinationId: string;
+
+  @Column
+  remoteKey: string;
+
+  @AllowNull(false)
+  @Default(ExportProcessor.defaultState)
+  @Column(DataType.ENUM(...ExportProcessorStates))
+  state: typeof ExportProcessorStates[number];
+
+  @Column
+  startedAt: Date;
+
+  @Column
+  processAt: Date;
+
+  @Default(0)
+  @Column
+  retryCount: number;
+
+  @Column
+  completedAt: Date;
+
+  @Column
+  errorMessage: string;
+
+  @HasMany(() => Export)
+  exports: Export[];
+
+  @Is("ofValidErrorLevel", (value) => {
+    if (value && !ERROR_LEVELS.includes(value)) {
+      throw new Error(`errorLevel must be one of: ${ERROR_LEVELS.join(",")}`);
+    }
+  })
+  @Column(DataType.ENUM(...ERROR_LEVELS))
+  errorLevel: ExportProcessorErrorLevel;
+
+  @BelongsTo(() => Destination)
+  destination: Destination;
+
+  async setError(error: Error, retryDelay: number = config.tasks.timeout) {
+    return ExportProcessorOps.setError(this, error, retryDelay);
+  }
+
+  async retry(retryDelay: number = config.tasks.timeout) {
+    return ExportProcessorOps.retry(this, retryDelay);
+  }
+
+  async complete() {
+    this.errorMessage = null;
+    this.errorLevel = null;
+    this.completedAt = new Date();
+    this.state = "complete";
+    await this.save();
+  }
+
+  async apiData(includeDestination = true) {
+    const exportsCount = await this.$count("exports");
+    const destination = await this.$get("destination", {
+      scope: null,
+    });
+
+    return {
+      id: this.id,
+      destination:
+        destination && includeDestination
+          ? await destination.apiData(false, false)
+          : null,
+      destinationName: destination ? destination.name : null,
+      state: this.state,
+      createdAt: APIData.formatDate(this.createdAt),
```

Comment:
Thanks for using APIData.formatDate - this helps SQLite deal with times a little better